### PR TITLE
Ability to show/hide menu bar

### PR DESCRIPTION
**Pull request by ghost**
Hi! Just added possibility to hide menu bar. In the View menu and F12 shortcut. Tested only in Linux though.

### CODE REVIEW

**Comment by esbrandt** [comment]
On Mac OSX 10.10.5 you can toggle the menu point, but nothing happens.

**Comment by daschuer** [comment]
Thank you for the PR. 

Please explain why the menu bar should go from you point of view. 
I know that GNOME is trying to get rid of any kind of menu, while Ubuntu patches the applications to get it back. 

I think it is a usability issue to hide the menu bar via mouse and need to remember that F12 is the magic key to bring it back. It should be more obvious to bring it back. It is too easy to look out. 

How about a "gear" menu like old nautilus or a header bar like the new one. Maybe it should be a skin and desktop environment option, if Mixxx has a classic menu bar or some usable replacement.

**Comment by dikouzmine** [comment]
Hi!

The main reason is that toolbar looks out of place, especially in fullscreen mode. Mixxx skins just don't look nice with this toolbar, it have custom interface that doesn't look good with standart qt widgets. Toolbar should be with customizable color maybe to look nice. Applying dark Qt theme is not a solution, because so many apps just made to be in light skins.

There's one dark skinned app, the Atom text editor from github. There they have also toggle menu bar option, and when menu bar is hidden you should pres Alt to reveal it again.

**Comment by uklotzde** [comment]
Hiding the option in full screen mode would free some unused vertical space, I like the idea! I also agree that in the Deere skin it looks misplaced (see screenshot). But it should be obvious on how to re-enable it without remembering some magic key.

deere full screen with menu bar

File: src/widget/wmainmenubar.cpp [hunk 1]
```diff
@@ -300,6 +300,22 @@ void WMainMenuBar::initialize() {
             pViewFullScreen, SLOT(setChecked(bool)));
     pViewMenu->addAction(pViewFullScreen);
 
+    QString toggleMenuBarTitle = tr("&Hide menu bar");
+    QString toggleMenuBarText = tr("Hide menu bar");
+    QAction* pToggleMenuBar = new QAction(toggleMenuBarTitle, this);
+    pToggleMenuBar->setShortcut(
+        QKeySequence(m_pKbdConfig->getValueString(ConfigKey("[KeyboardShortcuts]",
+                                                  "ViewMenu_ToggleMenuBar"),
+                                                  "F12")));
```
Comment (daschuer) [inline src/widget/wmainmenubar.cpp]: It is common in other Applications like Firefox or Nemo to show the menu bar if the ALT key is pressed.
Mixxx should do the same.

**Comment by daschuer** [comment]
In addition to the Alt Key, the Menu should be appear by a new Control Object.
This way the skin designer has the freedom to show the Menu, hide it because of an alternative solution or install a sensitive region on the very top of the skin, or a dedicated gear button. 
If this is in place, I have no objections to merge it.

**Comment by daschuer** [comment]
@dikouzmine: 
This could be implemented in the same way of [PreviewDeck]", "show_previewdeck"

https://github.com/mixxxdj/mixxx/blob/e59916caf72a256bb28b1722759a629c5cc9cf81/src/widget/wmainmenubar.cpp#L231

If the skin creates the control, the view menu is populated with the Action.

**Comment by Be-ing** [comment]
@dikouzmine are you interested in completing this by implementing a ControlObject as @daschuer proposed? I'd like to have this :)

**Comment by dikouzmine** [comment]
@Be-ing I'll take a deeper look at this task on weekend. It's not so straightforward. From what i see, WMainMenuBar is not like other Control Objects. It is tightly coupled with MixxxMainWindow and being created on startup. Lots of initialization code depends on it. I think i have to replace it with some other wrapper widget, and create menu bar as it's child when parsing a skin. I'll appreciate any hints and suggestions.

**Comment by Be-ing** [comment]
Glad you're still interested. @rryan , will this cause any issues with #941? Maybe it should wait for Mixxx 2.2?

**Comment by dikouzmine** [comment]
I've separated actions, signals and slots to separate invisible widget WMenuActionsContainer. Made MenuBar as widget controllable by skin. There is confusion with native menu bar (OS X, Unity), for now i just disabled it. Do we need it native? Also i'm having issue with new MenuBar widget: if i don't put it inside WidgetGroup or some other container, then visibility toggled on main skin container, but not on the MenuBar itself. There is also no straight way to make menu visibility toggle by Alt key, because QActions do not support this key alone. Might be possible to hack into qt message queue, but this solution don't fit into current framework so i don't know do we really need this?

**Comment by daschuer** [comment]
This branch does not merge cleanly with the current master. 

Maybe you should rebase you changes onto the current master. 
(Make a backup first) 

```bash
# if you have not already have an upstream branch  
git remote add upstream https://github.com/mixxxdj/mixxx.git
git fetch upstream 
git rebase --onto upstream/master 24c22123d25919624454e6aeb15164e60250f2b8
# resolve conflicts 
git mergetool 
git add <conflicting files>
git rebase --continue
# if you are satisfied
git push -f
```

**Comment by daschuer** [comment]
In general we should follow the Firefox way of handling the menu bar. 
In unity you cannot hid it, because it takes no extra space and is visible on mouse-over only. 
I think the situation is similar for Mac Os

**Comment by Be-ing** [comment]
This pull request has not been updated in over a year so I am closing it to reduce the clutter of open PRs. Also, I am not convinced that this is a great way to solve the problem of having space wasted by the menu bar. I think we should move everything that is only accessible through the menu bar currently into the main UI then remove the menu bar.

**Comment by daschuer** [comment]
This is IMHO a candidate for 2.1. We just need to rebase it and change the key to "Alt"

**Comment by Be-ing** [comment]
I am opposed to that approach. I find it annoying in Firefox, so much so that I had to search how to disable it because I accidentally hit Alt frequently. I have had Chrome users borrow my computer briefly and get confused by the sudden appearance of the menu bar.

**Comment by daschuer** [comment]
Are you against hiding the menu bar or against Alt as a shortcut?   

The request of hiding the menu bar comes up from time to time, so there is a valid demand. 
"Alt+..." is the key for all menu shortcuts so "Alt" should work intuitive.
And it is used in Firefox and Nemo.

**Comment by Be-ing** [comment]
I would prefer to remove the menu bar entirely. What is the need for it? Why not move the few items that are in the menu bar into the skins? Keeping the menu bar but putting the toggle for it buried in the preferences could work, but IMO there is no point in maintaining it.

**Comment by daschuer** [comment]
I prefer the menu bar over the skin menu. Because this way the user find the settings in a common way of all programs. The skin menu is hard to discover for a fist time user and its look changes from skin to skin.

**Comment by rryan** [comment]
FWIW menu bars help with accessibility. We do have some vision-impaired
users of Mixxx.

**Comment by Be-ing** [comment]
> this way the user find the settings in a common way of all programs.

This is not the case. More and more programs are adopting designs without menu bars. GNOME and Firefox have both abandoned menu bars.

> The skin menu is hard to discover for a fist time user

Really? Have you watched a new user struggle to find it? I think the cog icon is quite clearly related to settings. Zulip uses this and I have not seen anyone complain that they cannot find the preferences.

These are the only items in the menu bar currently that are not accessible another way:
  * Rescan Library: could be moved to the library area
  * Full screen: could be put in top section of skins next to feature show/hide buttons
  * Enable Keyboard Shortcuts: move to Interface pane of preferences?
  * Help menu: maybe move to a new menu that is shown when a "?" icon is clicked

**Comment by Be-ing** [comment]
> FWIW menu bars help with accessibility. We do have some vision-impaired users of Mixxx

I think it's a bit too broad of a statement to say that menu bars inherently help with accessibility. They are an accessible way to design a GUI, but that does not mean they are the only way. IIUC our custom skin system is really bad for accesibility.

**Comment by rryan** [comment]
> > FWIW menu bars help with accessibility. We do have some vision-impaired users of Mixxx
> 
> I think it's a bit too broad of a statement to say that menu bars inherently help with accessibility. They are an accessible way to design a GUI, but that does not mean they are the only way. IIUC our custom skin system is really bad for accesibility.

Specifically, what I mean is that the OS provides APIs to screen readers to inspect the menu tree. That won't be the case if we move all the options into the GUI (which at the moment is very hard to navigate for vision-impaired folks).

Anyway, good accessibility is often a benefit for non-impaired users too (e.g. curb-cuts). Having a uniform way to get to,  e.g. the preferences menu, that matches how other apps work is a benefit.

**Comment by Be-ing** [comment]
> Specifically, what I mean is that the OS provides APIs to screen readers to inspect the menu tree. That won't be the case if we move all the options into the GUI (which at the moment is very hard to navigate for vision-impaired folks).

What is the point of having a few menu items that do not control the music accessible to screen readers if the rest of the application is unusable for them? Those menu items do not obviate the need to make the rest of the UI accessible to screen readers. IIUC moving the few functions that are currently only available in the menu bar into the main GUI is not a step forward or backward for screen reader accessibility, it is a step sideways. If anyone wants to work on making Mixxx's skin system accessible, Qt has [accessibility documentation](https://doc.qt.io/qt-5.10/accessible.html).

**Comment by Be-ing** [comment]
Re-closing this as there has been no development, there are lots of merge conflicts, and there is no consensus that this should be merged. Let's move further discussion on this topic to Zulip instead of this old pull request.

**Pull request closed without merging**